### DISCUSSION TITLE
Screen Mirroring for iOS

### DISCUSSION THREAD

**Discussion by jagansekar-93** [appium/appium#20095] [Q&A]
Hi Folks, 

Is there something for iOS(iPhone and iPad) like Vysor, etc. Where we can screen mirror and control the device remotely?

This would helpful for managing the devices from remote. I have seen this options with DeviceFarm, headspin, browserstack, etc. But they are paid and enterprise owned. Appreciate your suggestions please

**Comment by jagansekar-93** [comment]
@jlipps @mykola-mokhnach @KazuCocoa - FYAP

**Reply by KazuCocoa** [reply]
I have no idea. This project focuses on automation, so remote control usage is out of scope. Such a usage may need optimization for the usage especially actions.

**Comment by vishwakarmaankushkumar82-droid** [comment]
Ok

**Comment by santtyamen** [comment]
Can you use Vysor 
https://www.vysor.io/

**Comment by nanoscopic** [comment]
Repeated screenshots through Appium can be used as a sort of faux video, and taps can be sent through Appium, but as said it isn't generally intended to do that. Doing that will result in a poor experience.

If you are on a Mac you can do this natively now through iPhone mirroring.

As far as the underlying mechanisms for how it works to make it work well:

Video:
1. Use a ReplayKit app to get and stream the video ( this is what MuTest does, as well as most device farm providers )
2. Use the USB video interface and stream video through that. The official MacOS video APIs can get video through it but it's still kind of flaky to get started. The underlying protocol is called Valeria. There are a few open source implementations floating around.
3. Via repeated screenshots. This is poor, but it is the only way currently supported by Appium to my knowledge.

Taps / Swipes:
1. XCUITest. Synthetic events can be done. "Live swiping" doesn't work through this unfortunately. Also the current app needs to be changed when the app is switched due to how XCUITest works. Appium supports this sort of...
2. Artificial bluetooth mouse. There are open source libraries floating around that give you a fake API controllable bluetooth mouse you can pair to your phone. This allows live swiping. There are a bunch of open source solutions floating around that work this way. The main limitation here is that it isn't absolute coordinates but is relative so dragging the mouse into / out of the phone can be wonky.
3. Universal control. Universal control has a full API for properly remote controlling the whole phone. I've reverse engineered it partially but there is still more work to be done. I don't know of any work besides my own to RE this. If someone is aware of any, please tell me.

Keyboard:
1. XCUITest via typeText. It works. Badly.
2. XCUITest via automating taps to the on screen keyboard. It works. Slowly, because Appium always fetches the elements again instead of just tapping precisely on the keys.
3. Mapping out the on screen keyboard then fast direct tapping on it via XCUITest. This works amazingly. Appium could be made to do this but I don't think anyone is attempting it. I've done it in my own software. I'm not aware of any open source currently that does this.
